### PR TITLE
[WIP] [C API] add multithreading test

### PR DESCRIPTION
**Pull request by panhaiqi**
### Details:
 - *[C API] Implement multithreading test cases*

### Tickets:
 - *94172*

### CODE REVIEW

File: src/bindings/c/tests/ov_multithreading_test.cpp [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+        ov_core_t* core = nullptr;
+        ov_core_create(&core);
+        ov_model_t* model = nullptr;
+        ov_core_read_model(core, xml_file_name.c_str(), bin_file_name.c_str(), &model);
+        ov_compiled_model_t* compiled_model = nullptr;
+        ov_core_compile_model(core, model, device_name.c_str(), 0, &compiled_model);
+    });
+}
+
+TEST_P(ov_multithreading_test, create_infer_request) {
```
Comment (ilyachur) [inline src/bindings/c/tests/ov_multithreading_test.cpp:52]: Maybe you need to cover the same cases like in cpp CoreThreading tests?

File: src/bindings/c/tests/ov_multithreading_test.cpp [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    std::vector<std::pair<std::string, std::string>> networks;
+    void set_up_networks(const unsigned int iterations = 100) {
+        for (unsigned i = 0; i < iterations; i++) {
+            std::pair<std::string, std::string> network(xml_file_name, bin_file_name);
+            networks.emplace_back(network);
+        }
+    }
+};
+
+INSTANTIATE_TEST_SUITE_P(device_name, ov_multithreading_test, ::testing::Values("CPU"));
```
Comment (riverlijunjie) [inline src/bindings/c/tests/ov_multithreading_test.cpp:49]:
```suggestion
INSTANTIATE_TEST_SUITE_P(ov_capi, ov_multithreading_test, ::testing::Values("CPU"));
```

File: src/bindings/c/tests/ov_multithreading_test.cpp [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+        OV_EXPECT_OK(ov_core_compile_model(core, model, device_name.c_str(), 0, &compiled_model));
+        EXPECT_NE(nullptr, compiled_model);
+
+        ov_compiled_model_free(compiled_model);
+        ov_model_free(model);
+        ov_core_free(core);
+    });
+}
+
+TEST_P(ov_multithreading_test, infer) {
```
Comment (riverlijunjie) [inline src/bindings/c/tests/ov_multithreading_test.cpp:103]: Can we also add async infer test case?

File: src/bindings/c/tests/ov_multithreading_test.cpp [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+        ov_capi_test_base::SetUp();
+    }
+
+    void TearDown() override {
+        ov_capi_test_base::TearDown();
+    }
+
+protected:
+    unsigned int iterations;
+    unsigned int threadsNum;
```
Comment (zhaixuejun1993) [inline src/bindings/c/tests/ov_multithreading_test.cpp:19]:
```suggestion
    unsigned int threads_num;
```
Pls keep the same naming style

File: src/bindings/c/tests/ov_multithreading_test.cpp [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+
+public:
+    void runParallel(std::function<void(void)> func,
```
Comment (zhaixuejun1993) [inline src/bindings/c/tests/ov_multithreading_test.cpp:22]: the same func_name()

**Comment by riverlijunjie** [comment]
@panhaiqi  could you run valgrind to check whether there is any memory leak in multiple threads test?

File: src/bindings/c/tests/ov_multithreading_test.cpp [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+
+    ov_compiled_model_t* compiled_model = nullptr;
+    OV_EXPECT_OK(ov_core_compile_model(core, model, device_name.c_str(), 0, &compiled_model));
+    EXPECT_NE(nullptr, compiled_model);
+
+    const char* key = ov_property_key_supported_properties;
+    char* result = nullptr;
+
+    run_parallel([&]() {
+        OV_EXPECT_OK(ov_compiled_model_get_property(compiled_model, key, &result));
```
Comment (riverlijunjie) [inline src/bindings/c/tests/ov_multithreading_test.cpp:65]: Memory leak is caused by this line. 
ov_compiled_model_get_property() will allocate array and return as result for each threads, but the result only be freed for one time.

Comment (riverlijunjie) [reply src/bindings/c/tests/ov_multithreading_test.cpp:65]:
```suggestion
        OV_EXPECT_OK(ov_compiled_model_get_property(compiled_model, key, &result));
        ov_free(result);
```

**Review by riverlijunjie — APPROVED** [review]
LGTM

File: src/bindings/c/tests/ov_multithreading_test.cpp [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+        ov_free(result);
+    });
+
+    ov_compiled_model_free(compiled_model);
+    ov_model_free(model);
+    ov_core_free(core);
+}
+
+TEST_P(ov_multithreading_test, compile_model) {
```
Comment (ilyachur) [inline src/bindings/c/tests/ov_multithreading_test.cpp:74]: Can we add separate tests for read_model and compile_model (without read)?

**Review by riverlijunjie — APPROVED** [review]

**Review by zhaixuejun1993 — APPROVED** [review]

**Comment by ilya-lavrenov** [comment]
why do we need these tests at all? why testing on C++ level is not enough?
does C API has some implementation details which can be non-thread safe?

**Pull request closed without merging**